packages/server/src/integrations/base/clientCache.ts

```diff
@@ -0,0 +1,65 @@
+import { Duration } from "@budibase/backend-core"
+import LRUCache from "lru-cache"
+
+interface Merkle {
+  hash: (json: Record<string, any>) => string
+  stringify: (json: Record<string, any>) => string
+}
+
+interface MerkleConstructor {
+  new (): Merkle
+}
+
+// no typescript definition - fill it in
+const { MerkleJson } = require("merkle-json") as {
+  MerkleJson: MerkleConstructor
+}
+
+const jsonHash = new MerkleJson()
+
+type ClientCacheValue<T> = {
+  destroy: (client: T) => Promise<void>
+  // can't really know the type of the client, they are all so different
+  client: T
+}
+
+const clientCache = new LRUCache<string, ClientCacheValue<any>>({
+  // max 50 active clients per service
+  max: 50,
+  ttl: Duration.fromMinutes(10).toMs(),
+  updateAgeOnGet: true,
+  dispose: async (value: ClientCacheValue<any>) => {
+    await value.destroy(value.client)
+  },
```

Comment:
When the max is reached, will an active connection get killed mid-query? My suspicion here is yes, which is possibly catastrophic. We could end up in situations where users are making requests, they're failing because they get killed by another request, they retry and kill another request, and it thrashes like this for a while.

I think there's probably a good reason that databases typically use pools and not caches for this sort of thing: you open a new identical connection if an existing one is already in use. Which makes me think, could we give out the same connection twice using this code? Is that bad?

Reply:
The first part is probably my biggest concern with this, I want to give this a test, but realistically it would be safer to mark that the connection is in-use and not close it until the job is finished - I think this is feasible.

The second part - we probably need to use a connection pool in the client - you are right that if a user makes multiple requests to the same DB these will all likely clash - I can adjust this to configure the client as a connection pool so that the cached version is just a pool of clients.

Reply:
I think caching a pool instead of a client solves all of the concerns here.

Reply:
Pooling implemented.